mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala

```diff
@@ -0,0 +1,136 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.feature
+
+import scala.util.parsing.combinator.RegexParsers
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.param.{Param, ParamMap}
+import org.apache.spark.ml.param.shared.{HasFeaturesCol, HasLabelCol}
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+
+/**
+ * :: Experimental ::
+ * Implements the transforms required for fitting a dataset against an R model formula. Currently
+ * we support a limited subset of the R operators, including '~' and '+'. Also see the R formula
+ * docs here: http://www.inside-r.org/r-doc/stats/formula
+ */
+@Experimental
+class RModelFormula(override val uid: String)
+  extends Transformer with HasFeaturesCol with HasLabelCol {
+
+  def this() = this(Identifiable.randomUID("rModelFormula"))
+
+  /**
+   * R formula parameter. The formula is provided in string form.
+   * @group setParam
+   */
+  val formula: Param[String] = new Param(this, "formula", "R model formula")
+
+  private var parsedFormula: Option[RFormula] = None
+
+  /**
+   * Sets the formula to use for this transformer. Must be called before use.
+   * @group setParam
+   * @param value an R formula in string form (e.g. "y ~ x + z")
+   */
+  def setFormula(value: String): this.type = {
+    parsedFormula = Some(RFormulaParser.parse(value))
+    set(formula, value)
+    this
+  }
+
+  /** @group getParam */
+  def getFormula: String = $(formula)
+
+  /** @group getParam */
+  def setFeaturesCol(col: String): this.type = set(featuresCol, col)
+
+  /** @group getParam */
+  def setLabelCol(col: String): this.type = set(labelCol, col)
+
+  override def transformSchema(schema: StructType): StructType = {
+    require(parsedFormula.isDefined, "Must call setFormula() first.")
+    val withFeatures = featureTransformer.transformSchema(schema)
+    val nullable = schema(parsedFormula.get.response).dataType match {
+      case _: NumericType | BooleanType => false
+      case _ => true
+    }
+    StructType(withFeatures.fields :+ StructField($(labelCol), DoubleType, nullable))
+  }
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    require(parsedFormula.isDefined, "Must call setFormula() first.")
+    transformLabel(featureTransformer.transform(dataset))
+  }
+
+  override def copy(extra: ParamMap): RModelFormula = defaultCopy(extra)
+
+  override def toString: String = s"RModelFormula(${get(formula)})"
+
+  private def transformLabel(dataset: DataFrame): DataFrame = {
+    val responseName = parsedFormula.get.response
+    dataset.schema(responseName).dataType match {
+      case _: NumericType | BooleanType =>
+        dataset.select(
+          col("*"),
+          dataset(responseName).cast(DoubleType).as($(labelCol)))
```

Comment:
What if the `responseName` is the same as `labelCol`? This may cause unexpected behavior. If the input is `DoubleType`, we should allow `labelCol` be the same as the target term in the formula. If we need to do transformation, then user should set a different `labelCol`. We can set the default `featuresCol` and `labelCol` based on the uid and hence it won't have name collision. I don't think this is a good solution, but I don't have good suggestions.

Btw, we can use `DataFrame.withColumn` to append a new column.